## DSBS

DSBS is a special tool to call variants and methylations for the DSBS project.
DSBS is written in Python3.5 (tested with v3.4 / v3.5 ).

## Installation
[python3+](https://www.python.org/)

[termcolor](https://pypi.python.org/pypi/termcolor/1.1.0)

[pysam](https://pypi.python.org/pypi/pysam)

[pyfasta](https://pypi.python.org/pypi/pyfasta/0.5.2)

## Usage
Once installed, you can use DSBS 
python3 DSBS.py 
'''
usage: DSBS.py [-h] [-v] [--maxDistance MAXDISTANCE] [--maxLen MAXLEN]
               [--mixLen MIXLEN] [--mixReadQual MIXREADQUAL]
               [--mixReadQualN MIXREADQUALN] [--maxN MAXN]
               [--maxSeqErr MAXSEQERR] [--maxSnp MAXSNP] [--maxIndel MAXINDEL]
               [--maxBp MAXBP] [--maxMut MAXMUT] [--minQual MINQUAL]
               [--minVaf MINVAF] [--secAlign] [--debug] [-q] [-c COVERAGE]
               [-p CPU] [-o OUTDIR] -g GENOMEFILE -d DBSNP [--Chr CHR]
               bamFile
DSBS.py: error: the following arguments are required: bamFile, -g/--genomeFile, -d/--dbsnp
'''

MultiQC is written in Python (tested with v2.7 / v3.4 / v3.5). It is
available on the [Python Package Index](https://pypi.python.org/pypi/multiqc/)
and through conda using [Bioconda](http://bioconda.github.io/).

## Contributions & Support

Contributions and suggestions for new features are welcome, as are bug reports! Please create a new [issue](https://github.com/tianguolangzi/DSBS/issues) for any of these, including example reports where possible.

If any question, please :
[@Zhang Kun](https://github.com/tianguolangzi) (tianguolangzi@yahoo.com; tianguolangzi@gmail.com)



### Contributors
Many thanks to those who have helped out with with project so far:
* Project lead and main author: [@Liang Jialong](https://github.com/tianguolangzi)
* Project lead and main author: [@Zhang Kun](https://github.com/llll)
